### PR TITLE
ci(PR): introduce new workflow PR

### LINKED ISSUE #2875: Trigger CI only when label A0,A2 for label conditions
### Problem to Solve

new problem after #2847, #2851 and #2874 

trigger CI when adding label works fine, but labeling any label ( even not A0, A2 ) can trigger the CI again as well

for example, 

1. when we are ready for review, we label our PR as A0, the CI start working
2. after sometime, somebody add a new label `gotissues`, the CI start rerunning ( this is not expected ) even though there isn't any new commits

### Notes

_No response_

### PR DESCRIPTION
**Pull request by clearloop**
Resolves #2875 
Squashed the commits #2878 based on the current master branch

```mermaid
---
title: Usages of check and build
---

flowchart LR
PR --> check{check}
CI --> check

CI --> |macos=true| build
PR -->|already-A0-A2, macos?| build{build}
Label -->|A0| build
Label -->|E2| build
```

+ [x] trigger build on new labels
  + [x] no more rerunning everything on changing labels
+ [x] make the workflow of PR independent

This PR contains only 4 commits with clear purposes just for making reviewing easier

| commit | description |
|-|-|
|6ee3af0| duplicate CI.yml that we can figure out the changes in the following commits |
|7703ed8| cherry-pick `check` and `check-fuzzer` steps from `CI.yaml` and create a new workflow `check.yml` |
|d3db8e5| cherry-pick `build`, `build-win-cross` and `build-macos`  steps from `CI.yaml` and create a new workflow `check.yml` |
|3806481| introduce `labels.yml` and `PR.yml` |
|d077d52| merge the cache implementation in #3001, embedded it status check of PR.yml, for CI.yml, it's at its previous place  |

@gear-tech/dev

### CODE REVIEW

**Comment by ukint-vs** [comment]
image

Why there is duplicate build actions?

File: .github/workflows/labels.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      - name: Cancel Previous Runs
+        uses: styfle/cancel-workflow-action@0.11.0
+        with:
+          access_token: ${{ github.token }}
+
+  build:
+    uses: ./.github/workflows/build.yml
+    # TODO: dispatch this workflow with github action API
+    # for more conditions. for example, if the there is
+    # already a build check exist. (issue-2888)
```
Comment (clearloop) [inline .github/workflows/labels.yml:21]: > Why there is duplicate build actions?

it is caused by the label `E2`, need to write more logic for avoiding it

**Review by ukint-vs — APPROVED** [review]

File: .github/workflows/labels.yml (hunk 1 above)
Comment (breathx) [reply .github/workflows/labels.yml:21]: But this duplicate actions in fact runs twice, isn't it bad?

Comment (clearloop) [reply .github/workflows/labels.yml:21]: or we can remove `labels.yml` in this PR, then introduce a complete version of it in another PR

Comment (breathx) [reply .github/workflows/labels.yml:21]: > or we can remove `labels.yml` in this PR, then introduce a complete version of it in another PR

that would be better

**Comment by StackOverflowExcept1on** [comment]
should it be duplicated here?
https://github.com/gear-tech/gear/blob/7560f8fc2230ef9c448a101c1ffe7ca577179be3/.github/workflows/CI.yaml#L44-L48
we have `- uses: ./.github/actions/prepare-sccache`

**Review by tltsutltsu — APPROVED** [review]

**Pull request merged**